### PR TITLE
Feature: add per-saltenv configuration parameter support

### LINKED ISSUE #29954: Per-saltenv gitfs configuration parameters
Hi,

I am trying to setup multiple environments (base, dev, prod) all getting their files from a git repository. I would like to be able to specify a root directory (or a branch) for each environment similar to ext_pillar. but in gitfs_remotes there seems to be no such option. for example:
 
    gitfs_remotes:
      - gitrepo:
        - root: base (# base directory in the repo contains base sls files)
        - env: base (# this option currently does not exists!)
      - gitrepo: ( #same repo above or can be another repo)
        - root: dev (# dev directory in the repo containing dev sls files)
        - env: dev (# branch or directory)

you get the idea.

There is only an option gitfs_base that allow me to specify which branch maps to the base environment. I would like to similarly map dev and prod to branches or directories.

One benefit of such structure is using the same state files to do different things in different environment such as managing different files:

manage_yum.sls

    /etc/yum.conf
      - file.managed:
        - source: salt://files/yum.conf

repo structure:

      /
      base/
        manage_yum.sls
        files/
          yum.conf
      dev/
        manage_yum.sls
        files/
          yum.conf

will copy different files based on the env its in.

### PR DESCRIPTION
**Pull request by terminalmage**
This pull-request adds per-remote configuration support to gitfs, allowing for
individual saltenvs to be manually mapped to branches/tags, as well as
per-saltenv mountpoints and roots.

Included also here are additions to the gitfs walkthrough which document usage
of this new feature (see [here](https://github.com/saltstack/salt/pull/33650/files#diff-6c68309414d1616a85add22eae559b6b)).

Resolves #29954.

### CODE REVIEW

**Comment by cachedout** [comment]
Go Go Jenkins!

File: salt/utils/__init__.py [hunk 1]
```diff
@@ -2563,7 +2562,11 @@ def is_dictlist(data):
     return False
 
 
-def repack_dictlist(data):
+def repack_dictlist(data,
```
Comment (jfindlay) [inline salt/utils/__init__.py]: @terminalmage, @ticosax, this change broke [`unit.states.dockerng_test.DockerngTestCase.test_running_discard_wrong_environment_values`](https://jenkins.saltstack.com/job/branch_tests/job/develop/job/salt-linode-cent6/1350/testReport/junit/unit.states.dockerng_test/DockerngTestCase/test_running_discard_wrong_environment_values/).  It is not clear to me what the best option is to fix the test, so would either of you mind looking into it?  Thanks.

Comment (terminalmage) [inline salt/utils/__init__.py]: I know what the problem is, here. Firstly, I did not code in a case where ``strict`` is actually used, and secondly it should default to ``False``, not ``True``, to preserve expected behavior with existing code.

Comment (terminalmage) [inline salt/utils/__init__.py]: See https://github.com/saltstack/salt/pull/33902

**Pull request merged**